packages/jmespath/src/Functions.ts

```diff
@@ -1,4 +1,605 @@
-// This is a placeholder for the real class. The actual implementation will be added in a subsequent PR.
-export class Functions {
-  public iAmAPlaceholder = true;
+import type {
+  JSONArray,
+  JSONObject,
+  JSONValue,
+} from '@aws-lambda-powertools/commons/types';
+import {
+  getType,
+  isNumber,
+  isRecord,
+} from '@aws-lambda-powertools/commons/typeutils';
+import type { Expression } from './Expression.js';
+import { JMESPathTypeError } from './errors.js';
+import type {
+  FunctionSignatureDecorator,
+  FunctionSignatureOptions,
+  JSONObject as JSONObjectType,
+} from './types.js';
+import { arityCheck, typeCheck } from './utils.js';
+
+/**
+ * A class that contains the built-in JMESPath functions.
+ *
+ * The built-in functions are implemented as methods on the Functions class.
+ * Each method is decorated with the `@Function.signature()` decorator to enforce the
+ * arity and types of the arguments passed to the function at runtime.
+ *
+ * You can extend the Functions class to add custom functions by creating a new class
+ * that extends Functions and adding new methods to it.
+ *
+ * @example
+ * ```typescript
+ * import { search } from '@aws-lambda-powertools/jmespath';
+ * import { Functions } from '@aws-lambda-powertools/jmespath/functions';
+ *
+ * class MyFunctions extends Functions {
+ *   ⁣@Functions.signature({
+ *     argumentsSpecs: [['number'], ['number']],
```

Comment:
shouldn't the argument spec be `array-number`?

Reply:
In this case, we are specifically asking for at least 1 and at most 2 arguments (`variadic: true`) both of which must be `number`. In the method we use `Array<number>` because we are getting the arguments (`args` plural) of the method.

If we put `array-number` then it'd be 1 single argument with type `Array<number>`.

For example, we mean this: `my_method(1, 2)` and not this `my_method([1, 2])`.